File: src/wdog/runner/runner.go
package main

import (
	"fmt"
	"encoding/json"
	"xqueue"
)

type Request struct {
	Event		string			`json:"event"`
	Args		map[string]string	`json:"args,omitempty"`
	ContentType	string			`json:"content,omitempty"`
	Body		string			`json:"body,omitempty"`
	Claims		map[string]interface{}	`json:"claims,omitempty"` // JWT
	Method		string			`json:"method,omitempty"`
	Path		string			`json:"path,omitempty"`

	B		*Body			`json:"-"`
}

type Responce struct {
	Status	int
}

type RunnerRes struct {
	Res	int
	Ret	string
	Status	int
}

func use(resp *Responce) {}

func main() {

	q, err := xqueue.OpenQueue("3")
	if err != nil {
		fmt.Printf("Can't open queue 3: %s", err.Error())
		return
	}

	for {
		var req Request

		err = q.Recv(&req)
		if err != nil {
			fmt.Printf("Can't receive message: %s", err.Error())
			return
		}

		if req.ContentType == "application/json" {
			var b Body

			err = json.Unmarshal([]byte(req.Body), &b)
			if err == nil {
				req.B = &b
			}
		}

		res, resp := Main(&req)

		var b []byte
		b, err = json.Marshal(res)
		if err != nil {
			fmt.Printf("Can't marshal the result: %s", err.Error())
			return
		}

		out := &RunnerRes { Res: 0, Ret: string(b) }

		if resp != nil {
			out.Status = resp.Status
		}

		err = q.Send(out)
		if err != nil {
			fmt.Printf("Can't send responce: %s", err.Error())
			return
		}
	}
}

File: src/wdog/runner/runner.swift
//
// © 2018 SwiftyCloud OÜ. All rights reserved.
// Info: info@swifty.cloud
//

import Foundation
import Glibc

public typealias Byte = UInt8

var qfd: Int32
qfd = 3

struct Request: Codable {
	var event: String?
	var args: [String:String]?
	var body: [Byte]?
	var claims: [String:String]?
	var request: String?
	var path: String?
}

struct Result: Codable {
	var res: Int
	var ret: String
	var status: Int
}

struct Response {
	var status: Int
	// The "then" thing is here
}

func load(data: [Byte]) -> Request {
	return try! JSONDecoder().decode(Request.self, from: Data(bytes: data))
}

func save(obj: Encodable, resp: Response?) -> Data {
	struct EncWrap: Encodable {
		let o: Encodable

		func encode(to encoder: Encoder) throws {
			try self.o.encode(to: encoder)
		}
	}

	let jstr = String(data: try! JSONEncoder().encode(EncWrap(o:obj)), encoding: .utf8)!
	let result = Result(res: 0, ret: jstr, status: resp?.status ?? 0)
	return try! JSONEncoder().encode(result)
}

while true {
	var msg = [Byte](repeating: 0x0, count: 1024)
	recv(qfd, &msg, 1024, 0)

	let rq = load(data: msg)
	var resp: Response?
	var ret: Encodable
	(ret, resp) = Main(rq: rq)
	var out = save(obj: ret, resp: resp)

	var pointer: UnsafePointer<UInt8>! = nil
	out.withUnsafeBytes { (u8Ptr: UnsafePointer<UInt8>) in
		pointer = u8Ptr
	}
	send(qfd, pointer, out.count, 0)
}
